libyul/optimiser/SSAReverser.cpp

```diff
@@ -0,0 +1,55 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <libyul/optimiser/SSAReverser.h>
+#include <libyul/AsmData.h>
+#include <libdevcore/CommonData.h>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+
+void SSAReverser::operator()(Block& _block)
+{
+	walkVector(_block.statements);
+	iterateReplacingMulti<2>(
+		_block.statements,
+		[&](Statement& _stmt1, Statement& _stmt2) -> boost::optional<vector<Statement>>
+		{
+			if (auto const* varDecl = boost::get<VariableDeclaration>(&_stmt1))
+				if (varDecl->variables.size() == 1 && varDecl->value)
+					if (auto const *assignment = boost::get<Assignment>(&_stmt2))
+						if (assignment->variableNames.size() == 1)
+							if (auto const* identifier = boost::get<Identifier>(assignment->value.get()))
+								if (identifier->name == varDecl->variables.front().name)
```

Comment:
The idea in writing ``iterateReplatingMulti`` was to avoid having to nest the ``if``s too deeply here... worked extremely well :-D...

Reply:
What about "return early"?
```
auto const* varDecl = boost::get<VariableDeclaration>(&_stmt1);
auto const *assignment = boost::get<Assignment>(&_stmt2);
if (!varDecl || !assignment)
  return {};
...
```

Reply:
I "flattened" it this way (it's a separate commit) - but actually I think the nested if's are in fact much easier on the eye - so personally I'd revert that commit and keep the nested ``if``s, but both is fine with me.

Reply:
As you prefer!

Reply:
Now that it's two cases I prefer it flattened, as it is :-).